akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -990,13 +1021,22 @@ private[stream] final class OrElse[T] extends GraphStage[UniformFanInShape[T, T]
 object GraphDSL extends GraphApply {
 
   class Builder[+M] private[stream] () {
+    private var unwiredIns = Set.empty[Inlet[_]]
+    private var unwiredOuts = Set.empty[Outlet[_]]
+
     private var traversalBuilderInProgress: TraversalBuilder = TraversalBuilder.empty()
 
     /**
      * INTERNAL API
      */
-    private[GraphDSL] def addEdge[T, U >: T](from: Outlet[T], to: Inlet[U]): Unit =
+    private[GraphDSL] def addEdge[T, U >: T](from: Outlet[T], to: Inlet[U]): Unit = {
+      require(unwiredIns.contains(to))
+      require(unwiredOuts.contains(from))
```

Comment:
I could change here easily to the `TraversalBuilder` methods but I also need the full set of unwired ports for the final check. Is it possible to get the full set of unwired Inlets and Outlets from the TraversalBuilder? Of course, that would be nicer, I agree.